Air Voice is a telecom company which is currently the leading telecom company in Europe. It currently deals with customer churn problem (Customer leaving the company). Build a churn prediction classification model. Exited is the target column (0 is stayed and 1 is exited).







In [2]:
import pandas as pd

In [3]:
av_telecom = pd.read_csv('F:\Skillenable\Data frames\Churn-Modelling.csv')
av_telecom.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [4]:
av_telecom.info()
av_telecom.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      10000 non-null  int64  
 1   Geography        10000 non-null  object 
 2   Gender           10000 non-null  object 
 3   Age              10000 non-null  int64  
 4   Tenure           10000 non-null  int64  
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   EstimatedSalary  10000 non-null  float64
 10  Exited           10000 non-null  int64  
dtypes: float64(2), int64(7), object(2)
memory usage: 859.5+ KB


CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [5]:
# crete dummy data for catagorical values

cat_col = ['Geography', 'Gender']
cat_col

['Geography', 'Gender']

In [6]:
av_telecom_dummy = pd.get_dummies(av_telecom, columns = cat_col, drop_first= True)
av_telecom_dummy.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Geography_Germany  \
0               1        101348.88       1                  0   
1               1        112542.58       0                  0   
2               0        113931.57       1                  0   
3               0         93826.63       0                  0   
4               1         79084.10       0                  0   

   Geography_Spain  Gender_Male  
0                0            0  
1                1            0  
2                0            0  
3                0            0  
4                1            0

In [7]:
# create input(X) and output(Y)
Y = av_telecom_dummy[['Exited']]
X = av_telecom_dummy.drop(columns=['Exited'])

# spliting data into test and train

from sklearn.model_selection import train_test_split

X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=47)
len(X_train),len(X_test),len(Y_train),len(Y_test)

(8000, 2000, 8000, 2000)

# Random forest

In [8]:
# creating random forest model
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=1000,random_state=47)
rf = model.fit(X_train,Y_train)


C:\Users\arun\AppData\Local\Temp/ipykernel_21280/2864290182.py:5: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf = model.fit(X_train,Y_train)


In [9]:
# Predicting the model on test data
Y_test


Exited
7432       0
5444       0
5448       1
8654       0
7543       0
...      ...
6746       0
2903       0
3893       0
7995       0
4070       0

[2000 rows x 1 columns]

In [10]:
Y_test['Pred_EXITED'] = model.predict(X_test)
Y_test

C:\Users\arun\AppData\Local\Temp/ipykernel_21280/2847273526.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Y_test['Pred_EXITED'] = model.predict(X_test)


Exited  Pred_EXITED
7432       0            0
5444       0            0
5448       1            1
8654       0            0
7543       0            0
...      ...          ...
6746       0            0
2903       0            0
3893       0            0
7995       0            0
4070       0            0

[2000 rows x 2 columns]

In [11]:
# Confusion metrics

from sklearn.metrics import confusion_matrix, accuracy_score

print(confusion_matrix(Y_test['Exited'],Y_test['Pred_EXITED']))

[[1511   66]
 [ 228  195]]


In [12]:
#Accuracy
print(accuracy_score(Y_test['Exited'],Y_test['Pred_EXITED']))


0.853


## 1) What is the accuracy of Random Forest Model?
## The accuracy of Random forest model for the given dataset is 85.3%

# Support Vector Machine (SVM)

In [13]:
# building SVM 
from sklearn.svm import SVC
model1 = SVC(random_state=47)

svc_model = model1.fit(X_train,Y_train)  # fit train data


C:\Users\arun\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [14]:
# prediction
Y_test['Pred_EXITED'] = svc_model.predict(X_test)
Y_test


C:\Users\arun\AppData\Local\Temp/ipykernel_21280/2877938925.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Y_test['Pred_EXITED'] = svc_model.predict(X_test)


Exited  Pred_EXITED
7432       0            0
5444       0            0
5448       1            0
8654       0            0
7543       0            0
...      ...          ...
6746       0            0
2903       0            0
3893       0            0
7995       0            0
4070       0            0

[2000 rows x 2 columns]

In [15]:
# confusion metrix
from sklearn.metrics import confusion_matrix,accuracy_score
print(confusion_matrix(Y_test['Exited'],Y_test['Pred_EXITED']))

[[1577    0]
 [ 423    0]]


In [17]:
# accuracy SVM
print(accuracy_score(Y_test['Exited'],Y_test['Pred_EXITED']))


0.7885


## 2) What is the accuracy of Support Vector Machine?
## The accuracy of SVM for the given dataset is 78.85%